Whitebackground

nam-van-nguyen opened the pull request:
Would you please merge my changes here: https://github.com/nam-van-nguyen/PianoBooster on the "whitebackground" branch?
I updated the code so that the UI preference dialog allowing people to set the red/green/blue integer code (0-255) for various colors.  Everything appears to work quite well.  In the future, it would help to add a color picker to help with the usability.

antartica committed:
Change background to white and adjust colors accordingly

nam-van-nguyen committed:
Add preference configurations for the play animation colors.
  + Added new fields in GUI preferences for various color fields
  + Moved the color settings over from CFG to CSettings, but keep the CFG methods still pointing to the CSettings due some dependency issue.
  + The color is 0-255 integer value for "red", "green" and "blue".
  + It would help if add a button for picking color, and even better if have a preview while changing the color.

nam-van-nguyen committed:
Added copyright statement to the new header file IColorPreference.h

nam-van-nguyen committed:
Updated to make the stave thickness
Reduced vertical bar thickness
Increased font size
Caching font object

nam-van-nguyen committed:
Added a color selector for each color setting so that user can click on the buttons to select a color

nam-van-nguyen committed:
Added an apply button to preference dialog so that users can view the change (while in listen mode) and make adjustment as needed.
  + The staves still don't update, needs to clear cache/redraw.

nam-van-nguyen committed:
Fixed first time loading not using default colors

nam-van-nguyen committed:
Fixed notes not shown when the note color changing is not set
Update background color when Apply button on the preference dialog is clicked
Update rendering of staves when Apply button is clicked on the preference dialog

pianobooster commented:
I would like to first run up and test your code before merging. However I have got other important things happening at moment, and so I will not be be able to look at this until the week after next. Sorry for the delay.

nam-van-nguyen committed:
Added background color for the color selector buttons.  The button's background only changes when clicked applied.

nam-van-nguyen committed:
Fixed played code code not working due to previous change, put back the previous code.

nam-van-nguyen commented:
> I would like to first run up and test your code before merging. However I have got other important things happening at moment, and so I will not be be able to look at this until the week after next. Sorry for the delay.

Nice to see your message.  This is not urgent, so whenever you have time would be great.  Thanks!

Martchus commented:
Thanks for implementing this. Not being able to configure colors is the limitation which annoyed me most. So I've just tested it out and it works generally.

One problem I've noticed is that some colors (e.g. the background) are not updated immediately after closing the preferences (only after restarting the app). Maybe that's also only because I'm using Wayland¹.

I like that the different colors are all customizable individually although I found the defaults not very nice. I suppose a completely white background and completely back notes would be better than grayish colors (for the sake of contrast and also for consistency with "normal" sheet music).

---

That's how it can look like after playing a little bit with the preferences:

![Screenshot_20220307_235338](https://user-images.githubusercontent.com/10248953/157131731-3520d5ef-e210-4790-93fd-7afffc7b6e59.png)

I suppose that's kind of pretty (well, it still isn't Musecore). Only the keyboard at the bottom looks a bit off with no lines to distinguish the keys.

---

I must also say that the preferences dialog looks quite noisy and could have some convenience feature. I'd definitely just put one "Colors:" label at the top instead of repeating it in every row within the list view.

Just to give you some inspiration, that's how I did once something similar in my project [Syncthing Tray](https://github.com/Martchus/syncthingtray) (the most relevant source code is actually in [qtutilities](https://github.com/Martchus/qtutilities/tree/master/paletteeditor)):

![Screenshot_20220308_001053](https://user-images.githubusercontent.com/10248953/157133726-2d38334c-3f73-425b-989b-230b71d33cd3.png)

And:
![Screenshot_20220308_001137](https://user-images.githubusercontent.com/10248953/157133823-53fa7bfe-0056-4f98-a8e7-aff87887827c.png)

*Not* having the RGB values directly in the list keeps the dialog much cleaner but of course it is a matter of taste. I also added a button to select some presets in my dialogs. Adding a small preview was actually also not *that* difficult and looks quite nice in my opinion.  I'm aware that the matrix-layout is likely not going to work for PianoBooster but some cells could be disabled if they don't apply. Note that the dialog from the 2nd screenshot allows one to pick a single color to compute everything else from it. That saves some time but still allows one to tweak details. Both dialogs also use a special "ColorButton" (also in the grid view) which looks quite nice in my opinion. The one in your dialog has some noisy looking rectangle inside (at least with Breeze style)².

Note that all of this is just for inspiration. I think the dialog is good enough to consider the PR mergable.

---

¹ Using KDE. To make PianoBooster draw *something* I had to run it via XWayland by passing the CLI args ` -platform xcb`.

² ![Screenshot_20220308_002643](https://user-images.githubusercontent.com/10248953/157135478-2ee7d0a7-f571-41cf-a187-e10d3ba43105.png)

nam-van-nguyen commented:
![preference](https://user-images.githubusercontent.com/642711/157157575-7db58879-9184-4ab7-90bb-09823b7a2aeb.png)
Here's how the image looks like on my computer.  The center of the buttons is a character.  Probably your system doesn't have the font.  I will look at making an image for it.

Although not as good as a preview, you can play a midi, then open the preference dialog, change some settings and click on the bottom left corner "Apply" button to see the change.

The background should change when the "Apply" button is clicked.  If not, see if you can get the latest code, or if you already did, then it's some peculiarity of Qt framework that may cause it, or the way you run it under KDE.

The noisy integer fields maybe removed completely.  It was written originally, so people can at least change the color using external color picker.  But I then added the color picker making the fields a bit redundant.

I was thinking about having presets also, or allow users to open/save settings or just color settings.  I don't know.  My time is very limited, and I don't know how much that is wanted.  A simple way is to just to post a list of them on GitHub as text, and then users just copy and paste to the "Piano Booster.ini" file.  I know it's not optimal, but at least we now have that ability.

The default color maybe not optimal.  I basically took what was already there (in the forked branch), and change the background a little since I stare at screen more than 10 hours a day, and the bright white background hurts my eyes.  Each has their own preference.

I appreciate you took time to review and providing feedback.  With limited time, I will see what I can do.

nam-van-nguyen commented:
It looks like the menu color and menu selected colors are not used anywhere.  Someone please confirm it so I can remove them.  I searched and didn't see any reference to them.

nam-van-nguyen committed:
Remove menu color and menu selected color since they are not used.

nam-van-nguyen committed:
Removed all color SpinBoxes and uses the color picker only.

nam-van-nguyen committed:
Fix crashes due to calling CSettings::loadSettings at an earlier time

nam-van-nguyen commented:
Here's an updated preference dialog:
![preference](https://user-images.githubusercontent.com/642711/157186167-b89b5f1a-307e-41c6-8525-25eb40f93ca7.png)

nam-van-nguyen committed:
Updated setting code to use cache for colors and also fix memory leak as the result

Martchus commented:
> The center of the buttons is a character.

It is a bit unfortunate that the character doesn't exist in KDE's default font and probably also not in other fonts. Maybe just keep the buttons empty as the simplest solution?

But I'd say the new dialog looks already much better.

> There appears to be segmentation fault after the program exits.

I also noticed that - so that's then *not* a problem specific to my setup. GDB (e.g. used via Qt Creator) tells one exactly where it happens so debugging it shouldn't be that hard.

I'll try your updated code later this evening.

Martchus commented:
Qt Creator should definitely allow you to do a debug build. At least for me it works like for any other CMake project. Not sure whether I'll be able to get to it again this evening but next time I can make a screenshot of my Qt Creator config if that's helpful. Of course you can also just invoke the build from the command line and use GDB directly (which has a built-in TUI) or use any other GDB fronted.

And no, using `exit` is not a correct solution. It simply means bypassing invocation of destructors but there's likely some UB in the code which needs fixing.

Martchus commented:
That's the drop-down menu you would add a build config:
![screenshot_20220308_201540](https://user-images.githubusercontent.com/10248953/157309101-4da1b2a6-9aae-42ac-97cf-c11cd79061d3.png)

Or you simply set `CMAKE_BUILD_TYPE` to `Debug` manually. I'd also recommend to use the latest version of Qt Creator because they recently improved the CMake integration *a lot*.

Martchus commented:
On my system the segfault happens in btw:

```
if (font!=nullptr) delete font;
```

Not the source of the issue, but I'd like to note that `delete` does a null-check on its own so the `if` is completely unnecessary.

Martchus commented:
Looks like font is an object of static storage duration that comes from here:

```
FTFont * setupFont() {
    static FTFont * cacheFont = NULL;
```

So calling `delete` on it later is bound to break.

Martchus commented on src/Draw.cpp:
```diff
@@ -37,36 +37,47 @@ typedef unsigned char guint8;
 int CDraw::m_forceCompileRedraw;
 
+FTFont * setupFont() {
+    static FTFont * cacheFont = NULL;
```
As mentioned in the other comment, other places `delete` this object expecting to take over ownership. However, only one of the possible callers can actually have ownership. Maybe the most elegant solution is an immediately invoked lambda:

```suggestion
    static FTFont cacheFont = [] {
        // compute params for FTBitmapFont constructor
        …
        // possibly do an early-exit
        …

        return FTBitmapFont(…);
    }();
```

Then you can just `return &cacheFont` and make sure all places using `setupFont()` do *not* take ownership.

nam-van-nguyen replied:
Thanks Martchus.  Using your suggestion of adding a debug build profile and adding an additional flag for debug, I was able to debug the code.  Good catch on that cached font.  It sure can cause issue.  I just removed the static flag for now.  I don't know how much overhead it may have, so at least with limited time, I am not going to play with it, and leave it for someone or you to take care of it.  Various issue like multi-threading, etc., I just don't want to touch beyond what I originally intended to do.  Thanks again for all the great feedback. I think it made a big difference.

pianobooster replied:
Guys before you get too excited about this. I think that themes are the way to go on this. So I would be very grateful if you all choose a nice set of colours that works well for you and a nice theme name and have a go adding the code to add themes. If we get a nice set of themes most users won't need to edit the colours themselves. To add a new theme with custom colours I favour defining themes in either a separate text config file or putting them in the standard pb config file. One of my aims of PB is to keep the UI really clean and simple so that even young children can use the program. Some young children (and even some adults) might not even know or understand what RGB is. The difficulty allowing unskilled users to control the colours is what happens if a user accidentally sets the background colour the same colour as the stave lines?

I am away at the moment and very busy so sorry I haven't been able to review anything yet. But I think we need themes implemented before we add any UI to add custom colours which I am undecided which is best custom colour UI or via a text file.

nam-van-nguyen committed:
Fixed application crash on exit due to deletion of a shared font object in Draw's destructor

nam-van-nguyen committed:
Added support for color themes
  + Added initial set of themes for color.
  + Additional theme files can be created and dropped into configuration folder
  + Each theme can be further customized to user's taste

nam-van-nguyen committed:
Minor change to Earth and Saturn themes

nam-van-nguyen committed:
Updates to various color themes

nam-van-nguyen committed:
Minor changes to Earth and Mars themes

nam-van-nguyen commented:
I made another branch for supporting background images.  I really like it, although I don't have enough time to adjust theme colors and finding images that would be good. 

I could possibly close this request and submit a pull request from that branch instead.

https://github.com/nam-van-nguyen/PianoBooster/tree/imagebackground
